server/sonar-server/src/main/java/org/sonar/server/measure/live/LiveMeasureComputerImpl.java

```diff
@@ -0,0 +1,146 @@
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.component.SnapshotDto;
+import org.sonar.server.computation.task.projectanalysis.qualitymodel.DebtRatingGrid;
+import org.sonar.server.computation.task.projectanalysis.qualitymodel.Rating;
+
+public class LiveMeasureComputerImpl implements LiveMeasureComputer {
+
+  private final DbClient dbClient;
+  private final MeasureMatrixLoader matrixLoader;
+  private final Configuration config;
+  private final IssueMetricFormulaFactory formulaFactory;
+
+  public LiveMeasureComputerImpl(DbClient dbClient, MeasureMatrixLoader matrixLoader, Configuration config, IssueMetricFormulaFactory formulaFactory) {
+    this.dbClient = dbClient;
+    this.matrixLoader = matrixLoader;
+    this.config = config;
+    this.formulaFactory = formulaFactory;
+  }
+
+  @Override
+  public void refresh(DbSession dbSession, Collection<ComponentDto> components) {
+    if (components.isEmpty()) {
+      return;
+    }
+
+    Map<String, List<ComponentDto>> componentsByProjectUuid = components.stream().collect(Collectors.groupingBy(ComponentDto::projectUuid));
+    for (List<ComponentDto> groupedComponents : componentsByProjectUuid.values()) {
+      refreshComponentsOnSameProject(dbSession, groupedComponents);
+    }
+  }
+
+  private void refreshComponentsOnSameProject(DbSession dbSession, List<ComponentDto> components) {
+    String projectUuid = components.iterator().next().projectUuid();
+    Optional<SnapshotDto> lastAnalysis = dbClient.snapshotDao().selectLastAnalysisByRootComponentUuid(dbSession, projectUuid);
+    if (!lastAnalysis.isPresent()) {
+      // project has been deleted at the same time ?
+      return;
+    }
+    Optional<Long> beginningOfLeakPeriod = lastAnalysis.map(SnapshotDto::getPeriodDate);
+
+    MeasureMatrix matrix = matrixLoader.load(dbSession, components, formulaFactory.getFormulaMetrics());
+    DebtRatingGrid debtRatingGrid = new DebtRatingGrid(config);
+
+    matrix.getBottomUpComponents().forEach(c -> {
+      IssueCounterImpl issueCounter = new IssueCounterImpl(dbClient.issueDao().selectGroupsOfComponentTreeOnLeak(dbSession, c, beginningOfLeakPeriod.orElse(Long.MAX_VALUE)));
+      FormulaContextImpl context = new FormulaContextImpl(matrix, debtRatingGrid);
+      for (IssueMetricFormula formula : formulaFactory.getFormulas()) {
+        // exclude leak formulas when leak period is not defined
+        if (beginningOfLeakPeriod.isPresent() || !formula.isOnLeak()) {
+          context.change(c, formula);
+          try {
+            formula.compute(context, issueCounter);
+          } catch (RuntimeException e) {
+            throw new IllegalStateException("Fail to compute " + formula.getMetric().getKey() + " on " + context.getComponent().getDbKey(), e);
+          }
+        }
+      }
+    });
+
+    // persist the measures that have been created or updated
+    // TODO test concurrency (CE deletes or updates the row)
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)